File: 2807-insert-greatest-common-divisors-in-linked-list/2807-insert-greatest-common-divisors-in-linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func gcd(a, b int) int {
	if b == 0 {
		return a
	}
	return gcd(b, a%b)
}
func insertGreatestCommonDivisors(head *ListNode) *ListNode {
	tmp := head
	for tmp.Next != nil {
		tmp.Next = &ListNode{gcd(tmp.Val, tmp.Next.Val), tmp.Next}
		tmp = tmp.Next.Next
	}
	return head
}

File: 2807-insert-greatest-common-divisors-in-linked-list/2807-insert-greatest-common-divisors-in-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int gcd(int a, int b) { return b == 0 ? a : gcd(b, a % b); }
    ListNode* insertGreatestCommonDivisors(ListNode* head) {
        ListNode* tmp = head;
        while (tmp->next) {
            tmp->next = new ListNode(gcd(tmp->val, tmp->next->val), tmp->next);
            tmp = tmp->next->next;
        }
        return head;
    }
};